group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1245,4 +1422,1304 @@ public static String consumerGroupSessionTimeoutKey(String groupId, String membe
+
+            timer.schedule(
+                genericGroupJoinKey(groupId),
+                group.rebalanceTimeoutMs(),
+                TimeUnit.MILLISECONDS,
+                false,
+                () -> completeGenericGroupJoin(group)
+            );
+
+            return EMPTY_RESULT;
+        } else {
+            group.initNextGeneration();
+            if (group.isInState(EMPTY)) {
+                log.info("Group {} with generation {} is now empty ({}-{})",
+                    groupId, group.generationId(), topicPartition.topic(), topicPartition.partition());
+
+                CompletableFuture<Void> appendFuture = new CompletableFuture<>();
+                appendFuture.whenComplete((__, t) -> {
+                    if (t != null) {
+                        // We failed to write the empty group metadata. If the broker fails before another rebalance,
+                        // the previous generation written to the log will become active again (and most likely timeout).
+                        // This should be safe since there are no active members in an empty generation, so we just warn.
+                        log.warn("Failed to write empty metadata for group {}: {}", group.groupId(), t.getMessage());
+                    }
+                });
+
+                List<Record> records = Collections.singletonList(
+                    RecordHelpers.newGroupMetadataRecord(group, metadataImage.features().metadataVersion())
+                );
+
+                return new CoordinatorResult<>(records, appendFuture);
+
+            } else {
+                log.info("Stabilized group {} generation {} ({}) with {} members",
+                    groupId, group.generationId(), topicPartition, group.size());
+
+                // Complete the awaiting join group response future for all the members after rebalancing
+                group.allMembers().forEach(member -> {
+                    List<JoinGroupResponseData.JoinGroupResponseMember> members = Collections.emptyList();
+                    if (group.isLeader(member.memberId())) {
+                        members = group.currentGenericGroupMembers();
+                    }
+
+                    JoinGroupResponseData response = new JoinGroupResponseData()
+                        .setMembers(members)
+                        .setMemberId(member.memberId())
+                        .setGenerationId(group.generationId())
+                        .setProtocolName(group.protocolName().orElse(null))
+                        .setProtocolType(group.protocolType().orElse(null))
+                        .setLeader(group.leaderOrNull())
+                        .setSkipAssignment(false)
+                        .setErrorCode(Errors.NONE.code());
+
+                    group.completeJoinFuture(member, response);
+                    timer.cancel(genericGroupHeartbeatKey(groupId, member.memberId()));
```

Comment:
@dajac i think our last discussion was to also revert this to the existing behavior right? i.e. not implement https://issues.apache.org/jira/browse/KAFKA-13766

Reply:
@jeffkbkim That's correct.

Reply:
So we actually need to reschedule the timer here, right?